src/terminal/parser/stateMachine.cpp

```diff
@@ -532,6 +539,63 @@ void StateMachine::_ActionParam(const wchar_t wch)
     }
 }
 
+// Routine Description:
+// - Triggers the SubParam action to indicate that the state machine should
+//   store this character as a part of a sub-parameter to a control sequence.
+// Arguments:
+// - wch - Character to dispatch.
+// Return Value:
+// - <none>
+void StateMachine::_ActionSubParam(const wchar_t wch)
+{
+    _trace.TraceOnAction(L"SubParam");
+
+    // Once we've reached the parameter limit, sub parameters are ignored.
+    if (!_parameterLimitOverflowed)
+    {
+        // If we have no parameters and we're about to add a sub parameter, add an empty parameter here.
+        if (_parameters.empty())
+        {
+            _parameters.push_back({});
+            const auto rangeStart = gsl::narrow<BYTE>(_subParameters.size());
+            _subParameterRanges.push_back({ rangeStart, rangeStart });
+        }
+
+        // On a delimiter, increase the number of sub params we've seen.
+        // "Empty" sub params should still count as a sub param -
+        //      eg "\x1b[0:::m" should be three sub params
+        if (_isSubParameterDelimiter(wch))
+        {
+            // If we receive a delimiter after we've already accumulated the
+            // maximum allowed sub parameters, then we need to set a flag to
+            // indicate that further sub parameter characters should be ignored.
+            // Similarly, we need to remove the last parameter so it isn't interpreted
+            // by the engine.
+            if (_subParameters.size() >= MAX_SUBPARAMETER_COUNT)
+            {
+                _parameterLimitOverflowed = true;
+                _parameters.pop_back();
+                _subParameterRanges.pop_back();
+            }

```

Comment:
I'm not sure that removing the last parameter here is the right approach. Because that means a sequence like this wouldn't work:

```
printf "\e[38:2::128:0:128:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1:1mMagenta?\e[m\n"
```

But this semicolon equivalent _does_ work, despite having the same limit:

```
printf "\e[38;2;128;0;128;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1;1mMagenta?\e[m\n"
```

Maybe it should only remove the last parameter if we haven't been able to add _any_ sub parameters yet? I guess that would be something like `_subParameterRanges.back().first >= MAX_SUBPARAMETER_COUNT`.

Reply:
This is addressed now. We simply parsed the above sequence as `"\e[38:2::128:0:128:1mMagenta?\e[m\n"` and this should be interpreted properly when we do support colon based Truecolor'*ing*.